contracts/strategies/direct-grants-simple/DirectGrantsSimpleStrategy.sol

```diff
@@ -0,0 +1,388 @@
+        if (upcomingMilestone[recipientId] != 0) {
+            revert RECIPIENT_ALREADY_ACCEPTED();
+        }
+
+        Recipient memory recipient = Recipient({
+            recipientAddress: recipientAddress,
+            isRegistryIdentity: registryGating ? true : isRegistryIdentity,
+            grantAmount: grantAmount,
+            metadata: metadata,
+            recipientStatus: InternalRecipientStatus.Pending
+        });
+
+        _recipients[recipientId] = recipient;
+
+        emit Registered(recipientId, _data, _sender);
+    }
+
+    /// @notice Allocate amount to recipent for direct grants
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the allocation
+    function _allocate(bytes memory _data, address _sender)
+        internal
+        virtual
+        override
+        nonReentrant
+        onlyPoolManager(_sender)
+    {
+        (address recipientId, RecipientStatus recipientStatus, uint256 grantAmount) =
+            abi.decode(_data, (address, RecipientStatus, uint256));
+
+        if (upcomingMilestone[recipientId] != 0) {
+            revert MILESTONES_ALREADY_SET();
+        }
+
+        Recipient storage recipient = _recipients[recipientId];
+
+        if (recipientStatus == RecipientStatus.Accepted) {
+            recipient.grantAmount = grantAmount;
+            recipient.recipientStatus = InternalRecipientStatus.Accepted;
+
+            IAllo.Pool memory pool = allo.getPool(poolId);
+            emit Allocated(recipientId, recipient.grantAmount, pool.token, _sender);
+        }
+    }
+
+    /// @notice Distribute the upcoming milestone
+    /// @param _sender The sender of the distribution
+    function _distribute(address[] memory _recipientIds, bytes memory, address _sender)
+        internal
+        virtual
+        override
+        onlyPoolManager(_sender)
+    {
+        uint256 recipientLength = _recipientIds.length;
+        for (uint256 i = 0; i < recipientLength;) {
+            _distributeUpcomingMilestone(_recipientIds[i], _sender);
+            unchecked {
+                i++;
+            }
+        }
+    }
+
+    function _distributeUpcomingMilestone(address _recipientId, address _sender) private {
+        uint256 milestoneToBeDistributed = upcomingMilestone[_recipientId];
+        Milestone[] storage recipientMilestones = milestones[_recipientId];
+
+        if (milestoneToBeDistributed > recipientMilestones.length) {
+            revert INVALID_MILESTONE();
+        }
+
+        Recipient memory recipient = _recipients[_recipientId];
+        Milestone storage milestone = recipientMilestones[milestoneToBeDistributed];
+
+        uint256 amount = recipient.grantAmount * milestone.amountPercentage / 1e18;
+
+        IAllo.Pool memory pool = allo.getPool(poolId);
+
+        allo.decreasePoolTotalFunding(poolId, amount);
+        _transferAmount(pool.token, recipient.recipientAddress, amount);
+
+        milestone.milestoneStatus = RecipientStatus.Accepted;
```

Comment:
should we also emit events when the milestone status changed?

Reply:
i think we should also should check if the milestone status is pending? - we don't want to payout rejected milestones by accident

Reply:
makes sense to do both

Reply:
```solidity
if (
            milestoneToBeDistributed > recipientMilestones.length
                || milestone.milestoneStatus == RecipientStatus.Rejected
        ) {
            revert INVALID_MILESTONE();
        }
```
if the milestone has NO status, you would still pay out the milstone
that's why it should `pending`. `pending` means it was submitted and not paid
suggested change:
```solidity
if (
            milestoneToBeDistributed > recipientMilestones.length
                || milestone.milestoneStatus != RecipientStatus.Pending
        ) {
            revert INVALID_MILESTONE();
        }
```